### PR TITLE
Course creator use case

### PR DESCRIPTION
**Pull request by mangelsnc**

### CODE REVIEW

File: applications/api/app/config/controllers/course.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+services:
+
+  codely.api.controller.course.post:
+    class: CodelyTv\Api\Controller\Course\CoursePostController
+    parent: codely.api.controller
```
Comment (JavierCane) [inline applications/api/app/config/controllers/course.yml]: New line at the end of files by default please 🙏

You can enable it in PhpStorm / IntelliJ:

preferences 2017-10-13 12-35-54

File: applications/api/app/config/controllers/_controllers.yml [hunk 2]
```diff
@@ -2,3 +2,4 @@ imports:
 
   - { resource: status.yml }
   - { resource: video.yml }
+  - { resource: course.yml }
```
Comment (JavierCane) [inline applications/api/app/config/controllers/_controllers.yml]: 👏👏👏

File: applications/api/src/Controller/Course/CoursePostController.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace CodelyTv\Api\Controller\Course;
+
+use CodelyTv\Api\Infrastructure\Controller\ApiController;
+use CodelyTv\Api\Infrastructure\Response\ApiHttpCreatedResponse;
+use CodelyTv\Context\Course\Module\Course\Application\Create\CreateCourseCommand;
+use CodelyTv\Types\ValueObject\Uuid;
+use Symfony\Component\HttpFoundation\Request;
+
+final class CoursePostController extends ApiController
```
Comment (JavierCane) [inline applications/api/src/Controller/Course/CoursePostController.php]: 👏👏👏

File: src/Context/Course/Module/Course/Application/Create/CourseCreator.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    public function __construct(CourseRepository $repository, DomainEventPublisher $publisher)
+    {
+        $this->repository = $repository;
+        $this->publisher = $publisher;
+    }
+
+    public function create(CourseId $id, CourseTitle $title, CourseDescription $description)
+    {
+        $course = new Course($id, $title, $description);
```
Comment (JavierCane) [inline src/Context/Course/Module/Course/Application/Create/CourseCreator.php]: Better to use a named constructor as in https://github.com/CodelyTV/cqrs-ddd-php-example/blob/d3da9074b765d0612e4c467e54d0cac94ba7d7c9/src/Context/Video/Module/Video/Application/Create/VideoCreator.php#L26

Why: https://codely.tv/screencasts/constructores-semanticos/ + record events inside the named constructor leaving the default constructor for reconstructing `Video` instances in a certain state without recording these events.

File: src/Context/Course/Module/Course/Application/Create/CreateCourseCommand.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+{
+    private $id;
+    private $title;
+    private $description;
+
+    public function __construct(Uuid $messageId, string $id, string $title, string $description)
+    {
+        parent::__construct($messageId);
+
+
```
Comment (JavierCane) [inline src/Context/Course/Module/Course/Application/Create/CreateCourseCommand.php]: unnecessary 2 blank lines 🙂

File: src/Context/Course/Module/Course/Application/Create/CreateCourseCommandHandler.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+final class CreateCourseCommandHandler
+{
+    private $creator;
+
+    public function __construct(CourseCreator $creator)
+    {
+        $this->creator = $creator;
+    }
+
+    public function __invoke(CreateCourseCommand $command)
```
Comment (JavierCane) [inline src/Context/Course/Module/Course/Application/Create/CreateCourseCommandHandler.php]: Add return type (`: void`) in order to make the side effect more explicit.

File: src/Context/Course/Module/Course/Domain/Course.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+        $course->record(
+            new CourseCreatedDomainEvent(
+                $id,
+                [
+                    'title' => $title->value(),
+                    'description' => $description->value(),
+                ]
+            )
+        );
```
Comment (JavierCane) [inline src/Context/Course/Module/Course/Domain/Course.php]: ⚠️ We should return the newly created instance (line 23)

+ We should use this named constructor from the `CourseCreator`

File: src/Context/Course/Module/Course/Domain/Course.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    private $description;
+
+    public function __construct(CourseId $id, CourseTitle $title, CourseDescription $description)
+    {
+        $this->id = $id;
+        $this->title = $title;
+        $this->description = $description;
+    }
+
+    public static function create(CourseId $id, CourseTitle $title, CourseDescription $description)
```
Comment (JavierCane) [inline src/Context/Course/Module/Course/Domain/Course.php]: Add return type (`: Course`)

File: src/Context/Course/Module/Course/Domain/CourseCreatedDomainEvent.php [hunk 9]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace CodelyTv\Context\Course\Module\Course\Domain;
+
+use CodelyTv\Shared\Domain\Bus\Event\DomainEvent;
+
+final class CourseCreatedDomainEvent extends DomainEvent
+{
+
```
Comment (JavierCane) [inline src/Context/Course/Module/Course/Domain/CourseCreatedDomainEvent.php]: Unnecessary blank line 🙂

File: src/Context/Course/Module/Course/Infraestructure/Persistence/CourseRepositoryMySql.php [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace CodelyTv\Context\Course\Module\Course\Infraestructure\Persistence;
+
+use CodelyTv\Context\Course\Module\Course\Domain\Course;
+use CodelyTv\Context\Course\Module\Course\Domain\CourseRepository;
+use CodelyTv\Infrastructure\Doctrine\Repository;
+
+final class CourseRepositoryMySql extends Repository implements CourseRepository
+{
+
```
Comment (JavierCane) [inline src/Context/Course/Module/Course/Infraestructure/Persistence/CourseRepositoryMySql.php]: Unnecessary blank line

File: src/Context/Course/Module/Course/Infraestructure/Persistence/CourseRepositoryMySql.php [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    public function save(Course $course)
```
Comment (JavierCane) [inline src/Context/Course/Module/Course/Infraestructure/Persistence/CourseRepositoryMySql.php]: Add return type (`: void`) in order to make the side effect more explicit.

File: src/Context/Course/Module/Course/Application/Create/CourseCreator.php [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+        $this->repository->save($course);
+
+        $this->publisher->publish(...$course->pullDomainEvents());
```
Comment (JavierCane) [inline src/Context/Course/Module/Course/Application/Create/CourseCreator.php]: 👌

File: src/Context/Course/Module/Course/Application/Create/CourseCreator.php [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    private $repository;
+    private $publisher;
+
+    public function __construct(CourseRepository $repository, DomainEventPublisher $publisher)
+    {
+        $this->repository = $repository;
+        $this->publisher = $publisher;
+    }
+
+    public function create(CourseId $id, CourseTitle $title, CourseDescription $description)
```
Comment (JavierCane) [inline src/Context/Course/Module/Course/Application/Create/CourseCreator.php]: Add return type (`: void`) in order to make the side effect more explicit.

File: src/Context/Course/Module/Course/Domain/Course.php (hunk 7 above)
Comment (mangelsnc) [reply src/Context/Course/Module/Course/Domain/Course.php]: 😱  I forgot the return!

File: applications/api/app/config/controllers/course.yml (hunk 1 above)
Comment (mangelsnc) [reply applications/api/app/config/controllers/course.yml]: Thanks for the tip!

**Comment by rgomezcasas** [comment]
@mangelsnc I merged it with a minor changes. Also I created an issue (https://github.com/CodelyTV/cqrs-ddd-php-example/issues/24) in order to add some missing things 😋

If you want to solve that issue feel free!! 😬

**Comment by rgomezcasas** [comment]
Also, @mangelsnc you're our first external contributor (https://github.com/CodelyTV/cqrs-ddd-php-example/graphs/contributors)!! 🎉🎉🎊🎊

**Comment by mangelsnc** [comment]
Cool! I will take a look at the issue! 😬

**Pull request merged**